auth/validator.go

```diff
@@ -5,119 +5,128 @@
 )
 
 // parseAndValidateJWT parses the token into its claims, verifies the claims and verifies the signature
-func (m *Middleware) parseAndValidateJWT(rawToken string) (*jwt.Token, error) {
-	token, parts, err := m.parser.ParseUnverified(rawToken, new(OIDCClaims))
+func (m *Middleware) parseAndValidateJWT(rawToken string) (Token, error) {
+
+	token, err := NewToken(rawToken)
 	if err != nil {
 		return nil, err
 	}
-	token.Signature = parts[2]
 
 	// get keyset
-	keySet, err := m.getOIDCTenant(token)
+	keySet, err := m.getOIDCTenant(token.Issuer())
 	if err != nil {
 		return nil, err
 	}
 
-	// verify claims
-	if err := m.validateClaims(token, keySet); err != nil {
-		return nil, err
-	}
-
 	// verify signature
-	if err = m.verifySignature(token, keySet); err != nil {
+	if err := m.verifySignature(token, *keySet); err != nil {
 		return nil, err
 	}
 
-	mapClaims, _, err := m.parser.ParseUnverified(rawToken, jwt.MapClaims{})
-	if err != nil {
+	// verify claims
+	if err := m.validateClaims(token, keySet); err != nil {
 		return nil, err
 	}
-	token.Claims.(*OIDCClaims).mapClaims = mapClaims.Claims.(jwt.MapClaims)
 
-	token.Valid = true
 	return token, nil
 }
 
-func (m *Middleware) verifySignature(t *jwt.Token, ks *oidcclient.OIDCTenant) error {
-	jwks, err := ks.GetJWKs()
+func (m *Middleware) verifySignature(t Token, keySet oidcclient.OIDCTenant) (err error) {
+	headers, err := getHeaders(t.GetTokenValue())
 	if err != nil {
-		return fmt.Errorf("token is unverifiable: failed to fetch token keys from remote: %v", err)
+		return err
 	}
-	if len(jwks) == 0 {
-		return fmt.Errorf("token is unverifiable: remote returned no jwk to verify the token")
+	kid := headers.KeyID()
+	alg := headers.Algorithm()
+
+	if alg == "" {
+		return errors.New("alg is missing from jwt header")
 	}
 
-	var jwk *oidcclient.JSONWebKey
+	publicKey, err := getPublicKey(kid, keySet)
+	if err != nil {
+		return err
+	}
+
+	//Parse and verify signature
+	_, err = jwt.ParseString(t.GetTokenValue(), jwt.WithVerify(alg, publicKey))
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	if kid := t.Header[propKeyID]; kid != nil {
+func getPublicKey(kid string, keySet oidcclient.OIDCTenant) (jwk.Key, error) {
+	jwks, _ := keySet.GetJWKs()
+	var jsonWebKey *oidcclient.JSONWebKey
+	if kid != "" {
 		for _, key := range jwks {
 			if key.Kid == kid {
-				jwk = key
+				jsonWebKey = key
 				break
 			}
 		}
-		if jwk == nil {
-			return fmt.Errorf("token is unverifiable: kid id specified in token not presented by remote")
+		if jsonWebKey == nil {
+			return nil, fmt.Errorf("token is unverifiable: 'kid' is specified in token, but no jwk provided by server")
 		}
 	} else if len(jwks) == 1 {
-		jwk = jwks[0]
+		jsonWebKey = jwks[0]
 	} else {
-		return fmt.Errorf("token is unverifiable: no kid specified in token and more than one verification key available")
+		return nil, fmt.Errorf("token is unverifiable: no kid specified in token and more than one jwk available from server")
```

Comment:
do you have a test case for this as well?

Reply:
no, this test would require another mockServer setup or refactoring jwk package to define interface that can be mocked.